consensus/istanbul/core/prepare.go

```diff
@@ -19,49 +19,124 @@ package core
 import (
 	"reflect"
 
+	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/consensus/istanbul"
 )
 
 func (c *core) sendPrepare() {
-	logger := c.logger.New("state", c.state)
+	logger := c.logger.New("state", c.state, "cur_round", c.current.Round(), "cur_seq", c.current.Sequence(), "func", "sendPrepare")
 
 	sub := c.current.Subject()
 	encodedSubject, err := Encode(sub)
 	if err != nil {
 		logger.Error("Failed to encode", "subject", sub)
 		return
 	}
-	c.broadcast(&message{
-		Code: msgPrepare,
+	logger.Trace("Sending prepare")
+	c.broadcast(&istanbul.Message{
+		Code: istanbul.MsgPrepare,
 		Msg:  encodedSubject,
 	})
 }
 
-func (c *core) handlePrepare(msg *message, src istanbul.Validator) error {
+func (c *core) handlePreparedCertificate(preparedCertificate istanbul.PreparedCertificate) error {
```

Comment:
Do we process the messages in the `preparedCertificate`? e.g. if somehow the preparedCertificate contained 2F+1 commit messages, would we then commit the block?

Reply:
No we don't. I don't think there is a case where the certificate should have 2F+1 commits, but even in that case, I don't think we lose liveness by not committing in that case.